dns-resolver-operator/src/bind.py

```diff
@@ -0,0 +1,225 @@
+                f"An exception occurred when reloading {constants.DNS_SNAP_NAME}. Reason: {e}"
+            )
+            logger.error(error_msg)
+            raise ReloadError(error_msg) from e
+
+    def start(self) -> None:
+        """Start the charmed-bind service.
+
+        Raises:
+            StartError: when encountering a SnapError
+        """
+        try:
+            cache = snap.SnapCache()
+            charmed_bind = cache[constants.DNS_SNAP_NAME]
+            charmed_bind.start()
+        except snap.SnapError as e:
+            error_msg = (
+                f"An exception occurred when stopping {constants.DNS_SNAP_NAME}. Reason: {e}"
+            )
+            logger.error(error_msg)
+            raise StartError(error_msg) from e
+
+    def stop(self) -> None:
+        """Stop the charmed-bind service.
+
+        Raises:
+            StopError: when encountering a SnapError
+        """
+        try:
+            cache = snap.SnapCache()
+            charmed_bind = cache[constants.DNS_SNAP_NAME]
+            charmed_bind.stop()
+        except snap.SnapError as e:
+            error_msg = (
+                f"An exception occurred when stopping {constants.DNS_SNAP_NAME}. Reason: {e}"
+            )
+            logger.error(error_msg)
+            raise StopError(error_msg) from e
+
+    def setup(self) -> None:
+        """Prepare the machine."""
+        self._install_snap_package(
+            snap_name=constants.DNS_SNAP_NAME,
+            snap_channel=constants.SNAP_PACKAGES[constants.DNS_SNAP_NAME]["channel"],
+        )
+        # We need to put the service zone in place so we call
+        # the following with an empty relation and topology.
+        self.update_config_and_reload()
+
+    def _write_file(self, path: pathlib.Path, content: str) -> None:
+        """Write a file to the filesystem.
+
+        This function exists to be easily mocked during unit tests.
+
+        Args:
+            path: path to the file
+            content: content of the file
+        """
+        pathlib.Path(path).write_text(
+            content,
+            encoding="utf-8",
+        )
+
+    def update_config_and_reload(
+        self, zones: list[str] | None = None, ips: list[str] | None = None
+    ) -> None:
+        """Update bind's config and reload bind.
+
+        Args:
+            zones: zones of the related authority servers
+            ips: ips of the related authority servers
+        """
+        if zones is None:
+            zones = []
+        if ips is None:
+            ips = []
+        start_time = time.time_ns()
+        logger.debug("Starting update of config")
+
+        # Write the service.test file
+        self._write_file(
+            pathlib.Path(constants.DNS_CONFIG_DIR) / f"db.{constants.ZONE_SERVICE_NAME}",
+            templates.ZONE_SERVICE.format(
+                serial=int(time.time() / 60),
+            ),
+        )
+
+        # Write the named.conf.local file
+        self._write_file(
```

Comment:
Should we check if the file exists before overwriting with empty zones/ips?

Reply:
I've added some defensive code in the reconcile loop with https://github.com/canonical/dns-operators/pull/201/commits/c279738f6a2dc1c0af44fe929e8bcd91e9d57b34